File: ios/Arculus/CSDK/WalletInitRecoverWallet.swift
import CSDK

final class WalletInitRecoverWallet: CSDKAPICall<Void> {
    private var nbrOfWords: Int
    
    init(wallet: OpaquePointer, nbrOfWords: Int) {
        self.nbrOfWords = nbrOfWords
        
        super.init(wallet: wallet)
    }
    
    override func request() async throws -> [Data] {
        var len: size_t = 0
        
        guard let bytes = WalletInitRecoverWalletRequest(wallet, nbrOfWords, &len) else {
            throw CSDKAPICallError(.requestCreationFailed("wallet init recover wallet"))
        }
        
        return [Data(bytes: bytes, count: len)]
    }
    
    override func response(bytes: [UInt8]) async throws -> ResponseType {
        let result = WalletInitRecoverWalletResponse(wallet, bytes, bytes.count)
        
        try CSDKAPICallError.validateWalletResponseCode(rc: result)
    }
}

File: android/src/main/java/com/aptoslabs/arculus/csdk/WalletInitRecoverWallet.kt
package com.aptoslabs.arculus.csdk

import com.sun.jna.Pointer

class WalletInitRecoverWallet(wallet: Pointer, private val nbrOfWords: Int) :
  CSDKAPICall<Unit>(wallet) {
  override suspend fun request(): Array<ByteArray?> {
    val len = SizeTByReference()

    val pointer = CSDK.WalletInitRecoverWalletRequest(wallet, nbrOfWords, len)

    if (pointer == Pointer.NULL) throw CSDKAPICallRequestCreationException()

    return arrayOf(pointer.getByteArray(0, len.toIntChecked()))
  }

  override suspend fun response(bytes: ByteArray) {
    val rc = CSDK.WalletInitRecoverWalletResponse(wallet, bytes, bytes.size)

    validateWalletResponse(rc)
  }
}
